### PR TITLE
Add interrupt node to x86 device tree

### PR DESCRIPTION
**Pull request by SavinayDharmappa**

### CODE REVIEW

File: dts/bindings/interrupt-controller/intel,ioapic.yaml [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+      description: compatible strings
+      constraint: "intel,ioapic"
+
+  - reg:
+      category: required
+      type: int
+      description: mmio register space
+      generation: define
+
+  - arm,num-irq-priority-bits:
```
Comment (galak) [inline dts/bindings/interrupt-controller/intel,ioapic.yaml]: remove this, looks like copy/paste error :)

File: dts/bindings/interrupt-controller/intel,ioapic.yaml [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+      category: required
+      type: int
+      description: number of bits of IRQ priorities
+      generation: define
+
+cell_string: IRQ
+
+"#cells":
+  - irq
+  - priority
```
Comment (galak) [inline dts/bindings/interrupt-controller/intel,ioapic.yaml]: it doesn't appear that the ioapic supports priority.  The second cell should be interrupt "sense":

  Device's interrupt property:

     interrupts = <P S>;

  The first number (P) represents the interrupt pin which is wired to the
  IO APIC. The second number (S) represents the sense of interrupt which
  should be configured and can be one of:
    1 - Edge Rising/High
    2 - Edge Falling/Low
    4 - Level High
    8 - Level Low

We should introduce a include/dt-bindings/interrupt-controller/irq.h with these defines.

**Review by galak — CHANGES_REQUESTED** [review]
some clean ups to yaml, and encoding of sense in IRQ property.

File: dts/bindings/interrupt-controller/intel,ioapic.yaml (hunk 2 above)
Comment (SavinayDharmappa) [reply dts/bindings/interrupt-controller/intel,ioapic.yaml]: ok.

**Comment by SavinayDharmappa** [comment]
shippable is failing as dts support for arduino_101_sss(arc part of arduino) is not yet added.

**Comment by SavinayDharmappa** [comment]
@galak  I have incorporated the changes. Please review it

File: dts/bindings/interrupt-controller/irq.h [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+
+#if defined(CONFIG_IOAPIC) || defined(CONFIG_MVIC)
+#define	LEVEL		0x00008000
+#define	EDGE		0x00000000
+#define	LOW		0x00002000
+#define	HIGH		0x00000000
+#define EDGE_RISING	0x00
+#define EDGE_FALLING	0x00002000
+#define LEVEL_HIGH	0x00008000
+#define LEVEL_LOW	0x0000A000
```
Comment (galak) [inline dts/bindings/interrupt-controller/irq.h]: Can we have this defined in terms of LEVEL, EDGE, LOW, HIGH above?

ie #define LEVEL_LOW (LEVEL | LOW)

File: dts/dts.cmake [hunk 4]
```diff
@@ -57,6 +57,7 @@ if(CONFIG_HAS_DTS)
     ${DTC_INCLUDE_FLAG_FOR_SOURCE}  # include the DTS source
     ${DTC_INCLUDE_FLAG_FOR_OVERLAY} # Possibly include an overlay after the source
     -I${PROJECT_SOURCE_DIR}/dts/common
+    -I${PROJECT_SOURCE_DIR}/dts/bindings/interrupt-controller
```
Comment (galak) [inline dts/dts.cmake]: this shouldn't be required.

The include should be like: #include <dt-bindings/i2c/i2c.h>

File: dts/bindings/interrupt-controller/irq.h [hunk 5]
```diff
@@ -0,0 +1,20 @@
+/*
```
Comment (galak) [inline dts/bindings/interrupt-controller/irq.h]: As these are IOAPIC/MVIC specific, this file should be:

intel-mvic.h & intel-ioapic.h (we can just have intel-mvic.h include intel-ioapic.h)

File: dts/x86/atom.dtsi [hunk 6]
```diff
@@ -1,4 +1,5 @@
 #include "skeleton.dtsi"
+#include "irq.h"
```
Comment (galak) [inline dts/x86/atom.dtsi]: #include <dt-bindings/interrupt-controller/intel-ioapic.h>

File: dts/x86/ia32.dtsi [hunk 7]
```diff
@@ -1,4 +1,5 @@
 #include "skeleton.dtsi"
+#include "irq.h"
```
Comment (galak) [inline dts/x86/ia32.dtsi]: #include <dt-bindings/interrupt-controller/intel-ioapic.h>

File: dts/x86/intel_curie.dtsi [hunk 8]
```diff
@@ -1,4 +1,5 @@
 #include "skeleton.dtsi"
+#include "irq.h"
```
Comment (galak) [inline dts/x86/intel_curie.dtsi]: #include <dt-bindings/interrupt-controller/intel-ioapic.h>

File: dts/x86/intel_quark_d2000.dtsi [hunk 9]
```diff
@@ -1,4 +1,5 @@
 #include "skeleton.dtsi"
+#include "irq.h"
```
Comment (galak) [inline dts/x86/intel_quark_d2000.dtsi]: #include <dt-bindings/interrupt-controller/intel-mvic.h>

File: dts/x86/quark_x1000.dtsi [hunk 10]
```diff
@@ -1,4 +1,5 @@
 #include "skeleton.dtsi"
+#include "irq.h"
```
Comment (galak) [inline dts/x86/quark_x1000.dtsi]: #include <dt-bindings/interrupt-controller/intel-ioapic.h>

**Review by galak — CHANGES_REQUESTED** [review]
Number of structural changes, nothing to major.

**Comment by galak** [comment]
Any updates?

File: dts/bindings/interrupt-controller/irq.h (hunk 3 above)
Comment (SavinayDharmappa) [reply dts/bindings/interrupt-controller/irq.h]: I had tried doing like that but i am getting parsing error . so i had done other way

**Comment by SavinayDharmappa** [comment]
@galak  i have updated the changes please review it. Following 104553d93f93dbf7892a0f3882f2b3f17215ad1b was breaking my patch after git bisect found it

File: include/dt-bindings/interrupt-controller/intel-ioapic.h [hunk 11]
```diff
@@ -0,0 +1,18 @@
...
+#define __DT_BINDING_IRQ_H
+
+#define	LEVEL		0x00008000
+#define	EDGE		0x00000000
+#define	LOW		0x00002000
+#define	HIGH		0x00000000
+#define EDGE_RISING	0x00
+#define EDGE_FALLING	0x00002000
+#define LEVEL_HIGH	0x00008000
+#define LEVEL_LOW	0x0000A000
```
Comment (galak) [inline include/dt-bindings/interrupt-controller/intel-ioapic.h]: Can EDGE_RISING, EDGE_FALLING, LEVEL_HIGH, LEVEL_LOW be defined like:

#define EDGE_RISING (EDGE | HIGH)
#define EDGE_FALLING (EDGE | LOW)

File: include/dt-bindings/interrupt-controller/intel-ioapic.h [hunk 12]
```diff
@@ -0,0 +1,18 @@
+/*
+ * Copyright (c) 2017 Linaro Limited
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#ifndef __DT_BINDING_IRQ_H
+#define __DT_BINDING_IRQ_H
+
+#define	LEVEL		0x00008000
```
Comment (galak) [inline include/dt-bindings/interrupt-controller/intel-ioapic.h]: We need to prefix these defines with IRQ_TYPE_:

LEVEL -> IRQ_TYPE_LEVEL
EDGE -> IRQ_TYPE_EDGE
...

**Review by galak — CHANGES_REQUESTED** [review]
A little bit of namespace / #define cleanup.  Otherwise looking good.

**Comment by SavinayDharmappa** [comment]
done the changes please review.

**Review by galak — CHANGES_REQUESTED** [review]
Need to update EDGE_RISING, EDGE_FALLING, LEVEL_HIGH, & LEVEL_LOW to be prefixed with IRQ_TYPE_

File: include/dt-bindings/interrupt-controller/intel-ioapic.h [hunk 13]
```diff
@@ -0,0 +1,18 @@
...
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#ifndef __DT_BINDING_IRQ_H
+#define __DT_BINDING_IRQ_H
+
+#define	IRQ_TYPE_LEVEL		0x00008000
+#define	IRQ_TYPE_EDGE		0x00000000
+#define	IRQ_TYPE_LOW		0x00002000
+#define	IRQ_TYPE_HIGH		0x00000000
+#define EDGE_RISING		(IRQ_TYPE_EDGE | IRQ_TYPE_HIGH)
```
Comment (galak) [inline include/dt-bindings/interrupt-controller/intel-ioapic.h]: all these defines should be prefixed with IRQ_TYPE_

Comment (SavinayDharmappa) [reply include/dt-bindings/interrupt-controller/intel-ioapic.h]: ok will do it

**Review by galak — APPROVED** [review]

**Pull request merged**